kobo/apps/trash_bin/utils.py

```diff
@@ -0,0 +1,378 @@
+from __future__ import annotations
+
+import json
+from copy import deepcopy
+from datetime import timedelta
+
+from django.contrib.auth import get_user_model
+from django.db import IntegrityError, models, transaction
+from django.db.models import F
+from django.db.models.signals import pre_delete
+from django.utils.timezone import now
+from django_celery_beat.models import (
+    ClockedSchedule,
+    PeriodicTask,
+    PeriodicTasks,
+)
+from rest_framework import status
+
+from kobo.apps.audit_log.models import AuditLog, AuditAction
+from kpi.exceptions import KobocatCommunicationError
+from kpi.models import Asset, ExportTask, ImportTask
+from kpi.utils.mongo_helper import MongoHelper
+from kpi.utils.storage import rmdir
+from .constants import DELETE_PROJECT_STR_PREFIX, DELETE_USER_STR_PREFIX
+from .exceptions import (
+    TrashIntegrityError,
+    TrashNotImplementedError,
+    TrashMongoDeleteOrphansError,
+    TrashTaskInProgressError,
+    TrashUnknownKobocatError,
+)
+from .models import TrashStatus
+from .models.account import AccountTrash
+from .models.project import ProjectTrash
+
+
+def delete_asset(request_author: 'auth.User', asset: 'kpi.Asset'):
+    deployment_backend_uuid = None
+    asset_id = asset.pk
+    asset_uid = asset.uid
+    owner_username = asset.owner.username
+    project_exports = []
+    if asset.has_deployment:
+        _delete_submissions(request_author, asset)
+        asset.deployment.delete()
+        project_exports = ExportTask.objects.filter(
+            data__source__endswith=f'/api/v2/assets/{asset.uid}/'
+        )
+
+    with transaction.atomic():
+        # Delete imports
+        ImportTask.objects.filter(
+            data__destination__endswith=f'/api/v2/assets/{asset.uid}/'
+        ).delete()
+        # Delete exports (and related files on storage)
+        for export in project_exports:
+            export.delete()
+
+        asset.delete()
+        AuditLog.objects.create(
+            app_label=asset._meta.app_label,
+            model_name=asset._meta.model_name,
+            object_id=asset_id,
+            user=request_author,
+            action=AuditAction.DELETE,
+            metadata={
+                'asset_uid': asset_uid,
+                'asset_name': asset.name,
+            }
+        )
+
+    # Delete media files left on storage
+    rmdir(f'{owner_username}/asset_files/{asset_uid}')
```

Comment:
excessive paranoia, but can we check that `asset_uid` is non-empty before calling `rmdir()`?